src/NServiceBus.Core/Transports/Msmq/MessagePump.cs

```diff
@@ -146,36 +146,41 @@ public void Start(PushRuntimeSettings limitations)
 
                     await concurrencyLimiter.WaitAsync(cancellationToken).ConfigureAwait(false);
 
-                    var tokenSource = new CancellationTokenSource();
-
                     var receiveTask = Task.Run(async () =>
                     {
-                        try
-                        {
-                            await receiveStrategy.ReceiveMessage(inputQueue, errorQueue, tokenSource, pipeline).ConfigureAwait(false);
-                            receiveCircuitBreaker.Success();
-                        }
-                        catch (MessageQueueException ex)
+                        using (var tokenSource = new CancellationTokenSource())
                         {
-                            if (ex.MessageQueueErrorCode == MessageQueueErrorCode.IOTimeout)
+                            try
                             {
-                                //We should only get an IOTimeout exception here if another process removed the message between us peeking and now.
-                                return;
+                                await receiveStrategy.ReceiveMessage(inputQueue, errorQueue, tokenSource, pipeline).ConfigureAwait(false);
+                                receiveCircuitBreaker.Success();
+                            }
+                            catch (MessageQueueException ex)
+                            {
+                                if (ex.MessageQueueErrorCode == MessageQueueErrorCode.IOTimeout)
+                                {
+                                    //We should only get an IOTimeout exception here if another process removed the message between us peeking and now.
+                                    return;
+                                }
+
+                                Logger.Warn("MSMQ receive operation failed", ex);
+                                await receiveCircuitBreaker.Failure(ex).ConfigureAwait(false);
+                            }
+                            catch (OperationCanceledException)
```

Comment:
We planned to expose in a later version the CancellationToke to the handlers over the context.